crates/recursor/src/recursor_dns_handle.rs

```diff
@@ -338,19 +339,32 @@ impl RecursorDnsHandle {
             }
         }
 
-        // collect missing IP addresses, select over them all, get the addresses
-        // make it configurable to query for all records?
+        // If we have no glue, collect missing IP addresses for non-child NS servers
+        // Querying for child NS servers can result in infinite recursion if the
+        // parent nameserver never returns glue records for child NS records, or does
+        // so based on cache freshness (observed with BIND)
+        //
+        // TODO: make it configurable to query for all records?
         if config_group.is_empty() && !need_ips_for_names.is_empty() {
-            debug!("need glue for {}", zone);
-            let a_resolves = need_ips_for_names.iter().take(1).map(|name| {
-                let a_query = Query::query(name.0.clone(), RecordType::A);
-                self.resolve(a_query, request_time, false).boxed()
-            });
+            debug!("need glue for {zone}");
 
-            let aaaa_resolves = need_ips_for_names.iter().take(1).map(|name| {
-                let aaaa_query = Query::query(name.0.clone(), RecordType::AAAA);
-                self.resolve(aaaa_query, request_time, false).boxed()
-            });
+            let a_resolves = need_ips_for_names
```

Comment:
These resolutions for `A` and `AAAA` are a bit repetitive. I think the original were a bit simpler, but with the added filters, I wonder if we should have a lambda or a generic function over the `RecordType` so that we can get a little more DRY? I'm just  a little concerned that if a bug is discovered here that we'll end up needing to change it across the 4 separate calls.

Reply:
I simplified this a bit; please take a look and let me know what you think.  The remaining duplication is down the differences in nameserver selection (child vs non-child NS) and using resolve in the non-child NS section and lookup in the child section and the resulting return type differences.

Reply:
Yes, I like that simplification, thank you!